airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java

```diff
@@ -427,6 +362,152 @@ private static Runnable getReplicationRunnable(final AirbyteSource source,
+        .withBytesEmitted(messageTracker.getTotalBytesEmitted())
+        .withSourceStateMessagesEmitted(messageTracker.getTotalSourceStateMessagesEmitted())
+        .withDestinationStateMessagesEmitted(messageTracker.getTotalDestinationStateMessagesEmitted())
+        .withMaxSecondsBeforeSourceStateMessageEmitted(messageTracker.getMaxSecondsToReceiveSourceStateMessage())
+        .withMeanSecondsBeforeSourceStateMessageEmitted(messageTracker.getMeanSecondsToReceiveSourceStateMessage())
+        .withMaxSecondsBetweenStateMessageEmittedandCommitted(messageTracker.getMaxSecondsBetweenStateMessageEmittedAndCommitted().orElse(null))
+        .withMeanSecondsBetweenStateMessageEmittedandCommitted(messageTracker.getMeanSecondsBetweenStateMessageEmittedAndCommitted().orElse(null))
+        .withReplicationStartTime(timeTracker.getReplicationStartTime())
+        .withReplicationEndTime(timeTracker.getReplicationEndTime())
+        .withSourceReadStartTime(timeTracker.getSourceReadStartTime())
+        .withSourceReadEndTime(timeTracker.getSourceReadEndTime())
+        .withDestinationWriteStartTime(timeTracker.getDestinationWriteStartTime())
+        .withDestinationWriteEndTime(timeTracker.getDestinationWriteEndTime());
+
+    if (outputStatus == ReplicationStatus.COMPLETED) {
+      totalSyncStats.setRecordsCommitted(totalSyncStats.getRecordsEmitted());
+    } else if (messageTracker.getTotalRecordsCommitted().isPresent()) {
+      totalSyncStats.setRecordsCommitted(messageTracker.getTotalRecordsCommitted().get());
+    } else {
+      LOGGER.warn("Could not reliably determine committed record counts, committed record stats will be set to null");
+      totalSyncStats.setRecordsCommitted(null);
+    }
+    return totalSyncStats;
+  }
+
+  private List<StreamSyncStats> getPerStreamStats(ReplicationStatus outputStatus) {
+    // assume every stream with stats is in streamToEmittedRecords map
+    return messageTracker.getStreamToEmittedRecords().keySet().stream().map(stream -> {
+      final SyncStats syncStats = new SyncStats()
+          .withRecordsEmitted(messageTracker.getStreamToEmittedRecords().get(stream))
+          .withBytesEmitted(messageTracker.getStreamToEmittedBytes().get(stream))
+          .withSourceStateMessagesEmitted(null)
+          .withDestinationStateMessagesEmitted(null);
+
+      if (outputStatus == ReplicationStatus.COMPLETED) {
+        syncStats.setRecordsCommitted(messageTracker.getStreamToEmittedRecords().get(stream));
+      } else if (messageTracker.getStreamToCommittedRecords().isPresent()) {
+        syncStats.setRecordsCommitted(messageTracker.getStreamToCommittedRecords().get().get(stream));
+      } else {
+        syncStats.setRecordsCommitted(null);
+      }
+      return new StreamSyncStats()
+          .withStreamName(stream)
+          .withStats(syncStats);
+    }).collect(Collectors.toList());
+  }
+
+  private void saveState(StandardSyncInput syncInput, ReplicationOutput output) {
```

Comment:
nit: `saveState` might be a bit misleading here, `output.withState` should be prepping the input for the next activity which should be `PersistStateActivity`.
I don't have a better name, at the moment, and I am hoping to be able to remove this soon, so it's probably not a big deal at the moment.

Reply:
good point! What about `prepStateForLaterSaving`? I can add a comment that this is done in the PersisStateActivity.

Reply:
Sounds good to me. Name looks a bit awkward but it is probably closer to the truth than saveState.

Reply:
accuracy over awkwardness!

Reply:
done!